rholang/src/main/scala/coop/rchain/rholang/interpreter/errors.scala

```diff
@@ -76,9 +76,10 @@ object errors {
       s"Top level free variables are not allowed: $freeVars."
   }
   final case class SubstituteError(override val toString: String) extends InterpreterError
-  final case class UnrecognizedInterpreterError(throwable: Throwable)
-      extends RuntimeException(throwable)
-      with InterpreterError
+  final case class UnrecognizedInterpreterError(throwable: Throwable) extends InterpreterError {
+    override def getMessage: String = throwable.getMessage
+    override def toString: String   = throwable.getLocalizedMessage
+  }
```

Comment:
This will discard the cause chain for the exception and cause debugging harder (not as hard as having the stackoverflow, addmittedly :)). I'd like us to find another way around this (that would retain the cause chain), or have a justification why its not possible.

Reply:
Also, adding a test showing that throwable.cause is set properly would be nice (I should've done that when doing my change, I guess). But I won't insist and will do that myself if you prefer, after this is merged.

Reply:
The stack trace you've pasted doesn't mention the `UnrecognizedInterpreterError` (which I believe it should) and doesn't illustrate that the whole cause chain would be included. Which I think would rely on the cause's toString method. It's not the case when cause is set properly.

`printStackTrace` relies on the cause chain only, and does some more logic than simply calling `toString` for the causes. The only way to have correct stack traces is to set the cause correctly.

You didn't provide a reason why changing that is not possible too.

Also, simply delegating all methods to the `cause` is a very misleading thing to do for a Throwable. Imagine e.g. how the debugger's output would look like.

Please change the `InterpreterError` to be an abstract sealed class, give it constructors with and without the `cause` (delegating to the Throwable's constructors appropriately), and use the one with `cause` for `UnrecognizedInterpreterError`.

Reply:
```
UnrecognizedInterpreterError(Unterminated string on line 14 begining at column 8)
	at coop.rchain.rholang.interpreter.Interpreter$.$anonfun$buildAST$2(Interpreter.scala:62)
	at cats.syntax.EitherOps$.leftMap$extension(either.scala:142)
	at coop.rchain.rholang.interpreter.Interpreter$.buildAST(Interpreter.scala:59)
	at coop.rchain.rholang.interpreter.Interpreter$.buildNormalizedTerm(Interpreter.scala:40)
	at coop.rchain.rholang.interpreter.RholangCLI$.processFile(RholangCLI.scala:120)
	at coop.rchain.rholang.interpreter.RholangCLI$.$anonfun$main$1(RholangCLI.scala:55)
	at coop.rchain.rholang.interpreter.RholangCLI$.$anonfun$main$1$adapted(RholangCLI.scala:55)
	at scala.collection.immutable.List.foreach(List.scala:389)
	at coop.rchain.rholang.interpreter.RholangCLI$.main(RholangCLI.scala:55)
	at coop.rchain.rholang.interpreter.RholangCLI.main(RholangCLI.scala)
```

is this better? It prints the _whole_ stacktrace for given error. Refactoring `InterpreterError` to become an `abstract class` is impossible because of different ways we construct subclasses of `InterpeterError` (from `Throwable`, plain string or contructor arguments ). This would require big engineering effort that is outside of the scope of this PR.

Reply:
It's better.

> It prints the whole stacktrace for given error.

for any error, even. Problem being, the stacktraces are incomplete because the cause field is not being set. Have a look at the implementation of printStackTrace to see how simplified and potentially incompatbile with tools this approach is. For once, any causes of the direct cause will not be printed.

```scala
new RuntimeException("a", new NullPointerException("b")).toString
//res0: String = java.lang.RuntimeException: a
```

> Refactoring InterpreterError to become an abstract class is impossible because of different ways we construct subclasses of InterpeterError (from Throwable, plain string or contructor arguments ). This would require big engineering effort that is outside of the scope of this PR.

I don't get what's the problem here. All non-cause parameters should be interpolated into a string and passed upstream as the `message` parameter for Throwable. I now see that we misuse `toString` (we should be using `message`, or a custom field/method), but we can always make toString delegate to getMessage. Then we should stop using Throwable.toString in a follow-up PR.